### PR TITLE
Testsuite: fixes for retail test

### PR DESCRIPTION
**Pull request by nadvornik**
## What does this PR change?

This fixes the setup of retail environment.

Before, the image-sync was called before the retail groups were created. As a result, the bug https://bugzilla.suse.com/show_bug.cgi?id=1199157 was not detected. This PR moves  image-sync to the right place.

Also, it adds the required restart after installation of saltboot formula and assigns the branch server to branch group.

The second commit increases the timeout for saltboot state to 600s. It took 400s at 
https://ci.suse.de/view/Manager/view/Manager-Head/job/manager-Head-dev-acceptance-tests-NUE/3237/

## GUI diff

No difference.

Before:

After:

- [ ] **DONE**

## Documentation
- No documentation needed: **add explanation. This can't be used if there is a GUI diff**
- No documentation needed: only internal and user invisible changes
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- No tests: **add explanation**
- No tests: already covered
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Review by aaannz — APPROVED** [review]

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature [hunk 1]
```diff
@@ -28,8 +28,12 @@ Feature: PXE boot a Retail terminal
     And I manually install the "vsftpd" formula on the server
     And I manually install the "saltboot" formula on the server
     And I manually install the "pxe" formula on the server
+    And I manually install the "image-sync" formula on the server
```
Comment (Bischoff) [inline testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: please remove the "pxe" and "image-sync" installations, they are not needed:
```
suma-bv-43-srv:~ # rpm -q patterns-suma_retail --requires 
bind-formula 
branch-network-formula 
dhcpd-formula 
image-sync-formula 
pxe-formula
```

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature [hunk 2]
```diff
@@ -28,8 +28,12 @@ Feature: PXE boot a Retail terminal
...
     And I synchronize all Salt dynamic modules on "proxy"
 
+  Scenario: Restart spacewalk services to apply saltboot reactor configuration
+    When I restart the spacewalk service
```
Comment (Bischoff) [inline testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:35]: Is this something the customer must apply, and is it documented?

If not, we should avoid such "cheating" because we know about the internals.

If this is to work around a bug, please document it as such:
```
# WORKAROUND: bsc#xxxxx
```

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature [hunk 3]
```diff
@@ -128,6 +132,10 @@ Feature: PXE boot a Retail terminal
     And I enter "Terminal branch: example.org" as "description"
     And I click on "Create Group"
     Then I should see a "System group example created." text
+    When I follow "Target Systems"
+    And I check the "proxy" client
+    And I click on "Add Systems"
+    Then I should see a "1 systems were added to example server group." text
```
Comment (Bischoff) [inline testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:138]: ok, I need to add it to build validation retail tests, i guess

Why do we need this group BTW?

Is it in our documentation?

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature [hunk 4]
```diff
@@ -191,7 +205,7 @@ Feature: PXE boot a Retail terminal
     And I wait until I see the name of "pxeboot_minion", refreshing the page
     And I follow this "pxeboot_minion" link
     # Workaround: Increase timeout temporarily get rid of timeout issues
-    And I wait at most 350 seconds until event "Apply states [util.syncstates, saltboot] scheduled by (none)" is completed
+    And I wait at most 600 seconds until event "Apply states [util.syncstates, saltboot] scheduled by (none)" is completed
```
Comment (Bischoff) [inline testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:208]: not fan, but ok

**Review by Bischoff — CHANGES_REQUESTED** [review]
see small changes request, and the few questions

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature (hunk 2 above)
Comment (nadvornik) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:35]: Documentation says that after installing retail pattern, reboot is required:
https://github.com/uyuni-project/uyuni-docs/blob/master/modules/retail/pages/retail-install-uyuni.adoc

This is equivalent to reboot.

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature (hunk 3 above)
Comment (nadvornik) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:138]: When an image is synced to particular branch server, a group pillar is created to indicate that the image is available for terminals in that branch.
There is a consistency check that the branch server must be part of the group, otherwise it generates an error message in the rhn log:
ERROR com.suse.manager.reactor.messaging.ImageSyncedEventMessageAction - Branch server uyuni-master-pxy.mgr.suse.de is not in group example

I thought it is in the documentation, but now I can't find it, I will take care of that.

Comment (nadvornik) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:138]: It is mentioned here https://github.com/uyuni-project/uyuni-docs/blob/master/modules/retail/pages/retail-components.adoc#branch-system-groups
However it is not clear that it must be added manually.

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature (hunk 1 above)
Comment (srbarrios) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: Ok, but my doubt is... where are we triggering the installation of that pattern in our CI proxy?

Comment (srbarrios) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: Shouldn't we use the steps described here? https://documentation.suse.com/suma/4.3/en/suse-manager/retail/retail-install-setup.html#_configure_services_on_the_branch_server

So... installing the formulas using `retail_branch_init` ?

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature (hunk 2 above)
Comment (srbarrios) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:35]: I don't see similar steps in https://documentation.suse.com/suma/4.3/en/suse-manager/retail and I'm confused.

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature (hunk 3 above)
Comment (Bischoff) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:138]: OK thanks for the details ; I applied same changes to BV retail tests

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature (hunk 1 above)
Comment (Bischoff) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: > Ok, but my doubt is... where are we triggering the installation of that pattern in our CI proxy?

here https://github.com/uyuni-project/uyuni/blob/363de44c8385addc31bc3fbb399d440d9375a329/testsuite/features/core/proxy_branch_network.feature#L31

I am more and more convinced that the order of scenarios in

testsuite/features/core/proxy_branch_network.feature
testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature

is completely wrong and needs refactoring. It already looks much better in BV Retail.

Also, the second file needs being split again to separate base retail and mass import scenarios.

Comment (Bischoff) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: > Shouldn't we use the steps described here? https://documentation.suse.com/suma/4.3/en/suse-manager/retail/retail-install-setup.html#_configure_services_on_the_branch_server
> 
> So... installing the formulas using `retail_branch_init` ?

I wrote those scenarios before that script got documented.

Comment (srbarrios) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: It was a good reason to merge these two features. https://github.com/SUSE/spacewalk/pull/9452

Comment (aaannz) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: I think this complete one scenario can be removed if we already have pattern installed. But then OTOH, why did @nadvornik feel the need to add image-sync here?

If we indeed need to install anything here, then I suggest to install pattern as @Bischoff does. Installing pattern is the supported and documented way (either explicitly or implicitly by SUMA Retail product).

Comment (nadvornik) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:31]: yes, I have already tested that all formulas are installed via pattern and this scenario can be removed

**Comment by nadvornik** [comment]
I have removed all formula installation from proxy_retail_pxeboot_and_mass_import.feature.
Installation of retail pattern in proxy_branch_network.feature should be enough.

File: testsuite/features/core/proxy_branch_network.feature [hunk 5]
```diff
@@ -35,6 +34,7 @@ Feature: Setup Uyuni for Retail branch network
     When I refresh the metadata for "server"
     When I install pattern "suma_retail" on this "server"
     And I wait for "patterns-suma_retail" to be installed on "server"
+    And I synchronize all Salt dynamic modules on "proxy"
```
Comment (vandabarata) [inline testsuite/features/core/proxy_branch_network.feature:37]: don't we also need this fix for `features/build_validation/retail/proxy_branch_network.feature`?

Comment (nadvornik) [reply testsuite/features/core/proxy_branch_network.feature:36]: it is already there

**Review by srbarrios — APPROVED** [review]
If you tested and it works, let's merge it.

File: testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature [hunk 6]
```diff
@@ -149,6 +146,12 @@ Feature: PXE boot a Retail terminal
     And I click on "Add Systems"
     Then I should see a "1 systems were added to SERVERS server group." text
 
+  Scenario: Move the image to the branch server
```
Comment (vandabarata) [inline testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:149]: Why was this moved? Is this image used or needed for the retail tests? it seems to be more related to the build image feature where it was before :thinking:

Comment (nadvornik) [reply testsuite/features/secondary/proxy_retail_pxeboot_and_mass_import.feature:149]: Because image-sync depends on the branch server group created in previous scenario of this feature.
(there is a compatibility fallback, so without the group it does not fail, but generate a warning message in the rhn log)
